### PR TITLE
Add matter opposition question to merits task list

### PR DESCRIPTION
**Pull request by cpjmcquillan**
When we ask the provider to provide information regarding the merit of the
application for legal aid, we ask them to complete the merits task list.

The merits task list is made up of sets of questions that apply to either the
application as a whole, or to the individual proceedings that make up the
application.

This adds a new application-level question to the merits task list.

The new question is applicable for all client-involvement types and is relevant
for any applications with at least one section 8 proceeding.

This will allow us to capture information from providers about why a given
matter is opposed by either the client or the opposing party (depending on
involvement type) in a free-text input.i

The questions in the merits task list are retrieved from the Legal Framework
API, and this adds a new flow step to the merits task loop to handle the
`why_matter_opposed` task.

When providers submit the form, matter opposed reasons are added to new matter
opposition records. These are associated to the application via a `has_one`
relationship, with the most recent record taking precedence.

The existing View <> Controller <> FormModel <> Model pattern in use across the
application has been adopted.

[Link to story](https://dsdmoj.atlassian.net/browse/AP-3490)

### CODE REVIEW

File: spec/requests/providers/proceeding_merits_task/attempts_to_settle_controller_spec.rb [hunk 1]
```diff
@@ -101,7 +101,7 @@
 
         it "updates the task list" do
           subject
-          expect(legal_aid_application.legal_framework_merits_task_list.serialized_data).to match(/name: :attempts_to_settle\n\s+dependencies: \*\d\n\s+state: :complete/)
+          expect(legal_aid_application.legal_framework_merits_task_list.serialized_data).to match(/name: :attempts_to_settle\n\s+dependencies: \*\d\d\n\s+state: :complete/)
```
Comment (colinbruce) [inline spec/requests/providers/proceeding_merits_task/attempts_to_settle_controller_spec.rb:104]: How do you feel about switching to `\d+`?  `\d\d` could be fragile if the sequences change in the future.
It's only recently become an issue as we started expanding beyond the original list of 8 questions, `\d+` matches on one or more digits?
```suggestion
          expect(legal_aid_application.legal_framework_merits_task_list.serialized_data).to match(/name: :attempts_to_settle\n\s+dependencies: \*\d+\n\s+state: :complete/)
```

**Review by colinbruce — APPROVED** [review]
LGTM, a couple of questions/observations but nothing blocking :+1

File: app/models/legal_aid_application.rb [hunk 2]
```diff
@@ -54,6 +54,7 @@ class LegalAidApplication < ApplicationRecord
   has_many :hmrc_responses, class_name: "HMRC::Response", dependent: :destroy, inverse_of: :legal_aid_application
   has_many :employments, dependent: :destroy
   has_many :regular_transactions, dependent: :destroy
+  has_one :matter_opposition, -> { order(created_at: :desc) }, class_name: "ApplicationMeritsTask::MatterOpposition", inverse_of: :legal_aid_application, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (cpjmcquillan) [reply app/models/legal_aid_application.rb:57]: I think `has_one` adds `LIMIT 1` to the SQL statement, which should be the order rows where added —but I find it helpful to be explicit with the scope to determine which record takes precedence.

In this case, we always want to the return the most recently created record.

File: spec/requests/providers/proceeding_merits_task/attempts_to_settle_controller_spec.rb (hunk 1 above)
Comment (cpjmcquillan) [reply spec/requests/providers/proceeding_merits_task/attempts_to_settle_controller_spec.rb:104]: Yeah that's perfect, thanks. 

I was just being lazy when making it go green.

It took me long enough to figure out why it was failing 😆

Comment (colinbruce) [reply spec/requests/providers/proceeding_merits_task/attempts_to_settle_controller_spec.rb:104]: Me too, the first time I hit it! 🤦‍♂️

File: app/models/legal_aid_application.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (colinbruce) [reply app/models/legal_aid_application.rb:57]: But if the controller loads an existing record shouldn't there only ever _be_ one? Or is this a belt and braces thing?

[the remark this answers is not in the document]
Comment (cpjmcquillan) [reply app/models/legal_aid_application.rb:57]: We don't ensure only one record can exist though at either the model or database level.

I actually assumed our use of `has_one` associations with the foreign key on the child table (i.e `matter_oppositions.legal_aid_application_id` vs `legal_aid_application.matter_opposition_id`) was intentional and designed to give us immutable records.

The current implementation just creates new matter opposition records, keeping the history in tact (if we ever wanted it).

If we don't want to use that approach, then I can add a unique index, validations and update the existing record.

What do you think?

[the remark this answers is not in the document]
Comment (colinbruce) [reply app/models/legal_aid_application.rb:57]: In most of these question controllers we use a pattern of 
```
def object
  @object ||= proceeding.object || proceeding.build_object
end
```

This should ensure that the existing object is loaded (if available) or created (if not)
I missed that omission from your controller, my bad!

[the remark this answers is not in the document]
Comment (colinbruce) [reply app/models/legal_aid_application.rb:57]: My brain is too all over the place this morning...
You _are_ doing that without the memoised prefix
Surely that should prevent multiple records per proceesding?

[the remark this answers is not in the document]
Comment (cpjmcquillan) [reply app/models/legal_aid_application.rb:57]: I'm not `merge(ing)_with_model`, so the `update` action initialises and saves a new record.

[the remark this answers is not in the document]
Comment (colinbruce) [reply app/models/legal_aid_application.rb:57]: Most of the ApplicationMeritsTask forms use merge_with_model, whats the advantage of your pattern and does it outweigh the cognitive load of being the outlier?

[the remark this answers is not in the document]
Comment (cpjmcquillan) [reply app/models/legal_aid_application.rb:57]: I don't think there's any particular advantage.

My initial assumptions aren't really relevant here. I don't think we're particularly interested in having immutable records or any form of event log —particularly only for this specific question.

I've updated the controller to follow the existing pattern and added a test to ensure existing records are updated.

**Pull request merged**